src/node/mod.rs

```diff
@@ -0,0 +1,1241 @@
+
+    fn handle_bootstrap_failure(&mut self) {
+        assert!(matches!(self.stage, Stage::Bootstrapping(_)));
+
+        info!("Failed to bootstrap. Terminating.");
+        let _ = self.user_event_tx.send(Event::Terminated);
+        self.stage = Stage::Terminated;
+    }
+
+    fn handle_connected_to(&mut self, _addr: SocketAddr) {}
+
+    fn handle_connection_failure(&mut self, addr: SocketAddr) {
+        if let Stage::Approved(stage) = &mut self.stage {
+            stage.handle_connection_failure(&mut self.core, addr);
+        } else {
+            trace!("ConnectionFailure from {}", addr);
+        }
+    }
+
+    fn handle_new_message(&mut self, sender: SocketAddr, bytes: Bytes) {
+        let msg = match MessageWithBytes::partial_from_bytes(bytes) {
+            Ok(msg) => msg,
+            Err(error) => {
+                debug!("Failed to deserialize message: {:?}", error);
+                return;
+            }
+        };
+
+        if let Err(error) = self.try_handle_message(Some(sender), msg) {
+            debug!("Failed to handle message: {:?}", error);
+        }
+    }
+
+    fn handle_unsent_message(&mut self, addr: SocketAddr, msg: Bytes, msg_token: Token) {
+        match self.core.handle_unsent_message(addr, msg, msg_token) {
+            PeerStatus::Normal => (),
+            PeerStatus::Lost => self.handle_peer_lost(addr),
+        }
+    }
+
+    fn handle_sent_message(&mut self, addr: SocketAddr, _msg: Bytes, token: Token) {
+        trace!("Successfully sent message with ID {} to {:?}", token, addr);
+        self.core.transport.target_succeeded(token, addr);
+    }
+
+    fn handle_timeout(&mut self, token: u64) {
+        if self.core.transport.handle_timeout(token) {
+            return;
+        }
+
+        match &mut self.stage {
+            Stage::Bootstrapping(stage) => stage.handle_timeout(&mut self.core, token),
+            Stage::Joining(stage) => {
+                if stage.handle_timeout(&mut self.core, token) {
+                    let network_cfg = stage.network_cfg;
+                    self.rebootstrap(network_cfg)
+                }
+            }
+            Stage::Approved(stage) => stage.handle_timeout(&mut self.core, token),
+            Stage::Terminated => {}
+        }
+    }
+
+    fn handle_peer_lost(&mut self, peer_addr: SocketAddr) {
+        if let Stage::Approved(stage) = &mut self.stage {
+            stage.handle_peer_lost(peer_addr);
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Message handling
+    ////////////////////////////////////////////////////////////////////////////
+
+    fn try_handle_message(
+        &mut self,
+        sender: Option<SocketAddr>,
+        mut msg_with_bytes: MessageWithBytes,
+    ) -> Result<()> {
+        trace!("try handle message {:?}", msg_with_bytes);
+
+        self.try_relay_message(sender, &msg_with_bytes)?;
+
+        if !self.in_dst_location(msg_with_bytes.message_dst()) {
```

Comment:
inside the above `try_relay_message` function call, there is a check of `self.in_dst_location`
can this double call to be avoided?
say make `try_relay_message` return with a bool (returns true if in dst), and do:
```rust
if !self.try_relay_message(sender, &msg_with_bytes)? {
  return Ok(());
}
```

Reply:
We could, in theory but I don't think it would be very clear. Also the `in_dst_location` call is very cheap, so it's not a problem to call it multiple times.